EP: MetaTransactions fixes

dorothy-zbornak opened the pull request:
## Description

- Adds a reentrancy guard to `executeMetaTransaction()` and `batchExecuteMetaTransactions()`.
  - This introduced new storage, `LibReentrancyGuardStorage`, and a new fixin `FixinReentrancyGuard`.
    - This is a flag-based reentrancy guard model, so we can have both distinct and overlapping reentrancy guards.
- Adds an ETH refund mechanism to meta-transactions.
- `transformERC20()` will now pass the `msg.sender` to transformers.
  - This is so FQT can refund any unused protocol fee to the sender, which may be distinct from the taker.
    - In the case of mtxs, the sender would be the EP itself, then the mtx refund mechanism would transfer it to the mtx relayer.
- Addresses internal feedback.
- Merged #2670 CD spot check feedback.
- Merged #2624 Fix `TransformerDeployer` so we can `kill()` transformers.
- Merged #2671 feature contract renames + minor cleanup.
- Rework nested revert error decoding to handle some edge cases.
- Fix shadowed variable in `TransformerDeployer`.
- Merged #2692 RFQT taker enforcement, kyber and mooni rollups, deployed addresses, etc.

## Testing instructions

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

merklejerk committed:
`@0x/contracts-zero-ex`: Fix `TransformerDeployer.kill()` calling the wrong `die()` interface.

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Add reentrancy guard to mtx functions
`@0x/contracts-zero-ex`: Add refund mechanism to mtxs
`@0x/contracts-zero-ex`: Pass sender to transfomers.
`@0x/contracts-zero-ex`: Refund protocol fees to `refundReceiver` in FQT.
`@0x/utils`: Add EP flavor of `IllegalReentrancyError`
`@0x/order-utils`: Add `refundReceiver` to FQT transform data.
`@0x/asset-swapper`: Add `refundReceiver` support to EP swap quote consumer.

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Add reentrancy guard to mtx functions
`@0x/contracts-zero-ex`: Add refund mechanism to mtxs
`@0x/contracts-zero-ex`: Pass sender to transfomers.
`@0x/contracts-zero-ex`: Refund protocol fees to `refundReceiver` in FQT.
`@0x/utils`: Add EP flavor of `IllegalReentrancyError`
`@0x/order-utils`: Add `refundReceiver` to FQT transform data.
`@0x/asset-swapper`: Add `refundReceiver` support to EP swap quote consumer.

hysz commented on contracts/zero-ex/contracts/src/features/SignatureValidator.sol:
```diff
@@ -37,6 +37,13 @@ contract SignatureValidator is
     using LibRichErrorsV06 for bytes;
 
+    /// @dev Exclusive upper limit on ECDSA signatures 'R' values.
+    ///      The valid range is given by fig (282) of the yellow paper.
```
+1 on ref!

hysz commented on contracts/zero-ex/contracts/src/features/SignatureValidator.sol:
```diff
@@ -160,12 +167,18 @@ contract SignatureValidator is
+            if (v < 27) {
+                // Handle clients that encode v as 0 or 1.
+                v += 27;
+            }
```
It looks like the valid values for `v` are `{27,28}`, after adjustment. Not sure if there's a strong security reason for verifying, but we could also assert this range.

dorothy-zbornak replied:
Anything outside that range should just cause `ecrecover()` to fail.

dorothy-zbornak replied:
I honestly think we should remove the `s` and `r` range checks. Signatures don't get included in any of our hashes, so replay isn't an issue. Plus, allowing `v` to be `0-1` or `27-28` introduces a new malleability since you can just take the same signature, flip `v` to `0-1`, and it'll be made valid.

hysz replied:
Oh interesting, that makes sense. Yeah I'm down to nix them. Perhaps just run by Paul real quick, since it was his recommendation.

dorothy-zbornak replied:
Paul's recco is to actually just drop the parity `v` support.

hysz commented on contracts/zero-ex/contracts/src/features/SignatureValidator.sol:
```diff
@@ -179,15 +192,21 @@ contract SignatureValidator is
+                // Handle clients that encode v as 0 or 1.
+                v += 27;
+            }
+            if (uint256(r) < ECDSA_SIGNATURE_R_LIMIT && uint256(s) < ECDSA_SIGNATURE_S_LIMIT) {
```
Same as above.

hysz commented on contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol:
```diff
@@ -249,6 +257,17 @@ contract FillQuoteTransformer is
         }
+
+        // Refund unspent protocol fees.
+        if (state.ethRemaining > 0 && data.refundReceiver != address(0)) {
```
Would it would indicate an internal error when `state.ethRemaining > 0 && data.refundReceiver  == 0`? Wondering if we should `revert` in that scenario.

dorothy-zbornak replied:
No. That's the behavior we have now. The `PayTakerTransformer` will funnel the unspent ETH to the taker.

hysz reviewed (approved):
Looks great, nice work! Left a couple small questions but looks g2g.

merklejerk committed:
`@0x/contracts-zero-ex`: Remove bugged parity signature compatibility.

merklejerk committed:
`@0x/contracts-zero-ex`: Remove bugged parity signature compatibility.

merklejerk committed:
`@0x/asset-swapper`: Fix rebase conflicts

merklejerk committed:
`@0x/asset-swapper`: Fix rebase conflicts

moodlezoup reviewed (approved):
good stuff, good tests, what's not to like?

merklejerk committed:
`@0x/contracts-integrations`: Add EP mtx tests.

merklejerk committed:
`@0x/order-utils`: Add `findTransformerNonce()` and `getTransformerAddress()` functions.

merklejerk committed:
`@0x/json-schemas`: Add `exchangeProxyMetaTransactionSchema`.

merklejerk committed:
`@0x/order-utils`: Fix EP signature utils schema assertion.

merklejerk committed:
`@0x/asset-swaper`: Use transformer utils from `@0x/order-utils`.

merklejerk committed:
`@0x/zero-ex`: Address CD post-audit feedback.
Add `LogMetadataTransformer`.

merklejerk committed:
`@0x/contracts-integrations`: Add EP mtx tests.

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Add `IUniswapV2Feature`.
`@0x/contracts-zero-ex`: Rename all feature contracts to have `Feature` suffix.
`@0x/contracts-zero-ex`: Return an `IZeroExContract` instance from `fullMigrateAsync()`.

dorothy-zbornak committed:
`@0x/migrations`: Update EP migration

dorothy-zbornak committed:
`@0x/asset-swapper`: Use `IZeroExContract` in EP swap quote consumer.

merklejerk committed:
Merge remote-tracking branch 'origin/fix/zero-ex/transformer-deployer-kill-fix' into fix/ep/meta-transactions

merklejerk committed:
Merge remote-tracking branch 'origin/feat/exchange-proxy/post-cd-audit' into fix/ep/meta-transactions

merklejerk committed:
`@0x/contracts-zero-ex`: Fix rebase artifacts.

dorothy-zbornak committed:
Merge branch 'feat/exchange-proxy/vip-uniswap' into fix/ep/meta-transactions

dorothy-zbornak committed:
`@0x/utils`: Fix nested revert decoding

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Remove UniswapFeature
`@0x/contracts-zero-ex`: Fix broken tests.

dorothy-zbornak committed:
`@0x/contracts-exchange`: Fix failing tests.

dorothy-zbornak committed:
`@0x/migrations`: Fix linter error.

dorothy-zbornak committed:
`@0x/utils`: Fix `StakingErrors.ExchangeManagerError` incorrect param name

dorothy-zbornak committed:
`@0x/contracts-integrations`: Remove stray `only` modifier.

dorothy-zbornak committed:
`@0x/contracts-integrations`: Fix broken tests.

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Fix shadowed variable in `TransformerDeployer`.

dorothy-zbornak committed:
update changelogs

dorothy-zbornak committed:
`@0x/utils`: Fix broken test

hysz reviewed (approved):
Updates LGTM!

moodlezoup reviewed (approved):
lgtm

dorothy-zbornak committed:
`@0x/contracts-zero-ex`: Add RFQT taker enforcement to FQT.
`@0x/contracts-zero-ex`: Remove redundant event from `BridgeAdapter`.
`@0x/contracts-zero-ex`: Treat all calldata as signed if quote signer is not in `TransformERC20Feature`.

`@0x/contracts-zero-ex`: Update bridge adapters

merklejerk committed:
`@0x/migrations`: Fix FQT deployment

merklejerk committed:
`@0x/order-utils`: Add `rfqtTakerAddress` to `FillQuoteTransformerData`

merklejerk committed:
`@0x/asset-swapper`: Add `rfqtTakerAddress` to `FillQuoteTransformerData`

moodlezoup commented on contracts/zero-ex/CHANGELOG.json:
```diff
@@ -1,4 +1,49 @@
+                "pr": 2657
+            },
+            {
+                "note": "Add `IUniswapV2Feature`",
```
no longer in this PR

dorothy-zbornak committed:
`@0x/contracts-integrations`: Add EP + MTX + RFQT integration tests.

merklejerk committed:
`@0x/asset-swapper`: Return Mooniswap pool in sampler and encode it in bridge data

merklejerk committed:
`@0x/contracts-zero-ex`: Decode secret mooniswap bridge data parameter.

merklejerk committed:
`@0x/contracts-zero-ex`: Remove `only` from tests.

merklejerk committed:
`@0x/asset-swapper`: Export `MooniswapFillData`.

merklejerk committed:
`@0x/contract-addresses`: Update transformer deployer and transformers for champagne-problems deployment.
`@0x/migrations`: Reorder transformer deployments.

merklejerk committed:
`@0x/asset-swapper`: Fix mooniswap encoding

merklejerk committed:
`@0x/contracts-zero-ex`: Remove `mooniswapRegistry` from `BridgeAdapter` addreses

merklejerk committed:
`@0x/migrations`: Remove `mooniswapRegistry` from `BridgeAdapter` addreses

merklejerk committed:
`@0x/contracts-zero-ex`: Fix Mooniswap WETH buying bug

merklejerk committed:
`@0x/asset-swapper`: Increase default error tolerance on quote simulation tests.
Run prettier.

merklejerk committed:
`@0x/asset-swapper`: Fix Mooniswap buy sampling bug.

merklejerk committed:
`@0x/contracts-zero-ex`: Only deploy `FlashWallet` in `TransformERC20` migration if it does not already exist.

merklejerk committed:
`@0x/asset-swapper`: Fix mooniswap sell sampling bug

dorothy-zbornak committed:
`@0x/contract-addresses`: Update EP addresses

dorothy-zbornak committed:
Merge pull request #2692 from 0xProject/feat/ep/fqt-rfqt-kyber-mooni-events

FQT: RFQT, more rollups, etc
